### PR TITLE
QuickJS engine: host-side, side-effect-free serialization via handles

### PR DESCRIPTION
**Pull request by TooTallNate**
## Summary

Moves the QuickJS engine's serialization **entirely to the host**, operating on `JSValueHandle`s — the serde bundle previously bundled by esbuild and evaluated inside the VM is gone. This mirrors the node:vm engine's architecture (the serializer is host code reaching into the sandbox realm) and is the QuickJS counterpart to #3257's side-effect-free serialization for node:vm.

Built on quickjs-wasi 3.3.0's host-side introspection primitives (`classId` brand checks, `identity`, descriptor reads, `vm.construct`, ephemeral functions; see vercel-labs/quickjs-wasi#24/vercel-labs/quickjs-wasi#26) and devalue's pluggable stringify/parse `operations` (already on `main` via #3257's devalue 5.9 bump).

## Performance

Counter-intuitively, host-side serde is **faster across every payload shape** — the boundary hops are cheap C calls, while the old approach ran the whole codec as interpreted JS inside an interpreted VM. Full-replay wall time (10-step payload-piping workflow, median of 7, details in [this comment](https://github.com/vercel/workflow/pull/3263#issuecomment-5184352617)):

| payload | in-VM bundle | host serde | speedup |
|---|---:|---:|---:|
| small (29 B) | 6.9 ms | 3.8 ms | 1.8× |
| 200-key object (7 KB) | 58.3 ms | 15.9 ms | 3.7× |
| 5k-object array (370 KB) | 4214.9 ms | 1447.8 ms | 2.9× |
| 512 KB string | 2494.9 ms | 23.8 ms | **105×** |
| Maps/Sets/Dates (10 KB) | 129.0 ms | 63.3 ms | 2.0× |

## Architecture

`runtime/quickjs-serde.ts` implements the workflow wire codec over handles:

- **Hybrid stringify value space**: workflow reducers return host shapes with handle leaves (exactly how the node:vm codec mixes host shapes with sandbox-realm leaves); every operation dispatches on `JSValueHandle` and falls back to `defaultStringifyOperations` for host values. Parse operations are handle-only — every revived value is built inside the VM through boot-captured constructors.
- **Side-effect freedom**: classification by engine brand (boot-sampled `classId` map, `isError`, `isProxy`) — never `instanceof` or `Symbol.toStringTag`; extraction through boot-captured intrinsics invoked with explicit receivers; property access through descriptors. Patched prototypes and spoofed brands can no longer perturb serialization (a `Symbol.toStringTag: 'Date'` spoof serializes as the plain object it is — the previous codec crashed on that input). The only guest code executed is what the contract always executed: `WORKFLOW_SERIALIZE`/`WORKFLOW_DESERIALIZE`, `__closureVarsFn`, `WORKFLOW_USE_STEP` on revival.
- **NUL (U+0000) safety**: `JS_ToCString`-backed extraction truncates at the first NUL and mangles NUL-bearing property keys. `guestString()` detects truncation against the handle's true guest length and recovers via in-VM `JSON.stringify` escaping; key enumeration verifies its fast path against a guest `Object.keys` count and re-extracts through length-aware key handles on mismatch; NUL-bearing keys route through handle-keyed get/set.
- **Bootstrap simplification**: pending ops carry RAW values (step input, hook metadata, abort payloads), and raw workflow results/errors; the host serializes through handles at collection time, with a per-VM byte cache shared between the suspension and terminal-drain paths (one byte sequence per op, even when getters are involved) and evicted as ops settle.
- **Determinism preserved**: correlationId ULIDs move host-side but draw from the **same seeded PRNG instance** as the VM's `Math.random`, so the interleaved draw sequence — and every generated ID — is identical to the in-VM factory's. `RetryableError`'s `retryAfter` fallback reads the guest's deterministic replay clock, not the host wall clock. Existing runs replay byte-for-byte.
- **Handle lifecycle**: serialize/deserialize passes sweep every intermediate handle created during the pass via `vm.withScope` (requires quickjs-wasi ≥ 3.3.1, whose borrowed-handle fix — vercel-labs/quickjs-wasi#31, found by this PR — makes scopes safe around host callbacks).

## Wire-format parity

Event logs persist across SDK versions, so parity with the previous in-VM codec is load-bearing (old runs must replay; node-engine steps must read VM-serialized inputs). The old codec's value-space implementation is retained as `serialization/workflow-vm.ts` (host reference codec) and a parity suite byte-compares against it in both directions — primitives/bigints/-0/NaN, containers, typed arrays/views/buffers, the full Error family with cause chains, shared refs + cycles, null-proto objects, boxed primitives, step-function proxies (closure vars + bound this), workflow refs, symbol-stamped stream handles, registry class instances, NUL-bearing strings/keys (including the truncate-collision enumeration shape), and patched-prototype/spoof resistance. Reducer/reviver key sets are pinned against `codec-devalue-vm`'s so drift fails loudly.

## Removed

- `scripts/build-vm-serde-bundle.js`, the generated `vm-serde-bundle.generated.ts`, `serialization/vm-bundle-entry.ts`, and the bundle eval in VM init.

## Testing

- 49 serde parity/safety tests (incl. the NUL suite); 24 quickjs-runtime tests green (fixtures built with the reference codec, so they validate parity end to end)
- Full `@workflow/core` suite: **1961 passed | 3 expected fail**
- Local e2e (`WORKFLOW_VM=quickjs`, dev server): `nullByteWorkflow` ✓, hooks 27✓, sleep/step races 3✓

## Follow-up candidates

- The VM-memory snapshotting stack (#3250/#3251) snapshots a VM that no longer contains serde state — restore paths should get simpler, but those branches need rebasing onto this change.
- ~~Upstream `handleHostCall` scope-tracking of C-owned argv pointers~~ — fixed in vercel-labs/quickjs-wasi#31, released in 3.3.1, and this PR now uses `withScope` directly. 3.3.1 also makes `memoryLimit` real on wasm32-wasi (vercel-labs/quickjs-wasi#33) and unregistered host callbacks throw (vercel-labs/quickjs-wasi#34).
- A bulk-free arena (flagged in quickjs-wasi#26) would still tighten the sweep costs for very large graphs.

### CODE REVIEW

File: packages/core/src/runtime/quickjs-runtime.ts [hunk 1]
```diff
@@ -998,18 +1004,27 @@ export async function startQuickJSWorkflow(
...
+  // draw sequence (and therefore every correlationId) is byte-identical to
+  // what the previous in-VM ULID factory produced for the same run. The
+  // time prefix is derived from the runId's embedded ULID (stable across
+  // invocations by construction — unlike `startedAt`, which differs
+  // between turbo's synthesized run object and the durably stored run),
+  // so two concurrent invocations of the same run produce IDENTICAL
+  // correlationIds and the world's EntityConflictError on `events.create`
+  // dedups one of each pair.
+  {
+    const ulidFactory = monotonicFactory(() => rng());
```
Comment (VaguelySerious) [inline packages/core/src/runtime/quickjs-runtime.ts]: ### AI Review: Blocking

Moving the ULID factory host-side is correct for this PR in isolation, but it makes the run's determinism state non-snapshottable, and #3251 (the sibling branch off the same base) snapshots the VM heap.

Two separate breakages once both land:

1. `__generateUlid` is installed via `newFunction` + `setProp` here. #3251's restore path re-registers only `random` and `__generateNanoid` (`quickjs-runtime.ts:1064` on that branch), so a restored heap calls a host callback that does not exist in the fresh process.
2. Adding it to that list is not sufficient. `monotonicFactory` with a fixed timestamp draws from the PRNG on its **first** call only and `incrementBase32`es afterwards, so all the sequencing state is in the host closure and is lost with the process. A rebuilt factory starts from a fresh random base. Locally the first post-restore correlation id came out `step_01JXT21Q004W1Z0086ZPBBFKHX` against `step_01JXT21Q001PKWKA2NK5BD215E` for the same step without a snapshot — divergent ids mean both `step_created` writes land and the step body executes twice.

The intrinsic capture has the mirror-image problem. `createQuickJSSerde(vm)` at `:988` is documented as capturing "before any user code runs", and only runs on the fresh-boot path. A restore path cannot satisfy that precondition by construction: the workflow bundle already evaluated in the invocation that produced the snapshot, so any serde created after a restore captures post-user-code intrinsics — which is exactly the tampering this PR exists to prevent.

Whichever of #3251 / #3263 lands second owns the fix. The structural version is for the VM-init code to return a descriptor of everything it installed (callbacks + captured intrinsics) so both entry paths are driven from one list and neither can silently diverge. A test that snapshots after step 1, restores, and asserts step 2's correlation id matches the unsnapshotted run would cover it.

File: pnpm-workspace.yaml [hunk 2]
```diff
@@ -50,6 +50,7 @@ minimumReleaseAge: 2880
 
 minimumReleaseAgeExclude:
   - '@turbo/*'
+  - 'devalue'
```
Comment (VaguelySerious) [inline pnpm-workspace.yaml]: AI Review: Note

This permanently exempts `devalue` from the 2880-minute cooldown, which is a lasting change to the supply-chain posture for a package that is now on the critical path of every serialized step argument. The existing entries are first-party or vendor-scoped (`@workflow/*`, `@vercel/*`, `turbo`); `devalue` is a third-party dependency and does not fit that pattern.

The motivation is presumably that 5.9.0 is too new to install. #3257 bumps the same override to 5.9.0 **without** the exclusion — it just waits out the cooldown. That is the cleaner precedent, and it avoids a conflict: both PRs edit the same `devalue:` line in `overrides`, so whichever lands second has to resolve it, and if it is this one the exclusion silently sticks around forever.

If the cooldown genuinely needs to be bypassed to unblock CI, a comment naming the reason and the version makes it removable later; as written there is nothing to tell a future reader it was meant to be temporary.

**Review by VaguelySerious — COMMENTED** [review]
AI review: blocking issues found

File: packages/core/src/runtime/quickjs-serde.ts [hunk 3]
```diff
@@ -0,0 +1,1985 @@
...
+      streamName,
+      hookToken,
+      aborted,
+      reason: aborted ? reason : undefined,
+    };
+  };
+
+  const reducers: Record<string, (value: any) => any> = {
+    // Order is wire-significant (first match wins) and mirrors
+    // codec-devalue-vm.ts getReducersForMode('workflow') exactly.
```
Comment (VaguelySerious) [inline packages/core/src/runtime/quickjs-serde.ts:865]: AI Review: Note

"matches `codec-devalue-vm.ts` `getReducersForMode('workflow')` exactly" is asserted only by this comment. Nothing fails if the two drift — a reducer added to the shared set will silently not exist on the QuickJS path, and the symptom is a value that round-trips as a plain object instead of its class.

This is the same concern I raised on #3048 about `common-vm.ts`, and deleting the in-VM bundle plus the byte-parity suite in `quickjs-serde.test.ts` genuinely improves it — a parity test is much better than a hand-maintained duplicate. What is still missing is exhaustiveness. I checked the two key sets at this ref and they currently match 43/43, so a test comparing `Object.keys` of the reducer tables (rather than comparing serialized bytes for a fixed value list) would lock that in and fail loudly on the next addition, which is the failure mode you want.

File: packages/core/src/runtime/quickjs-runtime.ts [hunk 4]
```diff
@@ -1291,11 +1321,11 @@ async function processEvents(
...
             });
-            const bytesHandle = vm.newUint8Array(decryptedOutput);
-            vm.setProp(vm.global, '__tmp_result', bytesHandle);
-            bytesHandle.dispose();
+            const valueHandle = serde.deserialize(decryptedOutput);
+            vm.setProp(vm.global, '__tmp_result', valueHandle);
+            valueHandle.dispose();
             vm.evalCode(
-              `globalThis.__resolvers["${escapedCid}"].resolve(globalThis[Symbol.for("workflow-deserialize")](globalThis.__tmp_result));` +
+              `globalThis.__resolvers["${escapedCid}"].resolve(globalThis.__tmp_result);` +
```
Comment (VaguelySerious) [inline packages/core/src/runtime/quickjs-runtime.ts]: AI Review: Nit

`escapedCid` still escapes only `"`, and the result is interpolated into source that gets `evalCode`d. Same nit as on #3048 — it survives the refactor. `JSON.stringify(cid)` handles quotes, backslashes and control characters in one call and removes the need to reason about what a correlation id can contain.

File: packages/core/src/runtime/quickjs-runtime.ts (hunk 1 above)
Comment (TooTallNate) [reply packages/core/src/runtime/quickjs-runtime.ts]: Acknowledged — this is the cross-branch collision you also flagged on #3251, and the reconciliation is now concrete rather than aspirational. #3251 (7f4d5174c) reworked snapshot determinism to position-based PRNG fast-forward: `session.snapshot()` returns the draw count, it's persisted as `SnapshotMetadata.rngDraws`, and restore re-seeds from the BASE seed and fast-forwards — which is exactly the state a host-side monotonic ULID factory rebuilds from, since it draws through the same counted PRNG (your point 2's fresh-random-base problem disappears when the rebuilt factory's first draw happens at the recorded position). #3251 also centralized host callbacks into ONE list driving both boot and restore, so when this branch restacks over the snapshot stack, `__generateUlid` joins that list and can't skip re-registration (your point 1). The serde intrinsic-capture point on restore is the remaining design question and I've kept your framing: a restore path can't honor 'before any user code runs', so the restacked serde will need to capture at snapshot time (host-side state, serializable) or accept post-bundle capture with the tamper-window documented — that lands with the restack, not here, since the branches are still parallel.

File: pnpm-workspace.yaml (hunk 2 above)
Comment (TooTallNate) [reply pnpm-workspace.yaml]: Fixed in a68205135 — the `devalue` entry is removed from `minimumReleaseAgeExclude` entirely. You (and the bot) were right that it was unnecessary: with the exact version pinned via the catalog + lockfile, both frozen and regular installs succeed without the waiver, so the cooldown protection stays intact and the #3257 conflict resolves to a no-op on this line.

File: packages/core/src/runtime/quickjs-serde.ts (hunk 3 above)
Comment (TooTallNate) [reply packages/core/src/runtime/quickjs-serde.ts:865]: Fixed in a68205135 — took your exact suggestion: `QuickJSSerde` now exposes `reducerKeys`/`reviverKeys`, `codec-devalue-vm` exports its workflow-mode key sets, and two tests pin them to each other (reducers compared in ORDER, since first-match-wins makes order wire-significant; revivers as sets). 43/43 keys match at this ref, and the next addition to either side fails loudly.

File: packages/core/src/runtime/quickjs-runtime.ts (hunk 4 above)
Comment (TooTallNate) [reply packages/core/src/runtime/quickjs-runtime.ts]: Fixed via the base branch: #3048 (c0b2f77e8) replaced all hand-rolled quote-only escaping with `JSON.stringify(cid)` (and `JSON.stringify(opType)`), and this branch is now restacked on top of it (a68205135), so the refactored sites inherit the hardening — no `replace(/\"/g, ...)` remains in either file.

File: packages/core/src/runtime/quickjs-serde.ts [hunk 5]
```diff
@@ -0,0 +1,1996 @@
...
+    return identity;
+  };
+
+  const primitiveOf = (handle: JSValueHandle): unknown => {
+    if (handle.isUndefined) return undefined;
+    if (handle.isNull) return null;
+    if (handle.isBool) return handle.toBoolean();
+    if (handle.isNumber) return handle.toNumber();
+    if (handle.isBigInt) return guestBigInt(handle);
+    return handle.toString();
```
Comment (pranaygp) [inline packages/core/src/runtime/quickjs-serde.ts]: **Blocker (CI-proven, all 15 quickjs legs): `handle.toString()` routes through `JS_ToCString`, which is NUL-terminated — any guest string containing U+0000 arrives silently truncated.** This is the funnel for essentially every string crossing the boundary: primitives here, property keys, symbol descriptions, RegExp source, Headers/URL values. Worse, **NUL-bearing object keys are silently *dropped*** — the enumeration APIs (`keys()`/`getOwnPropertyKeys()`/`propertyIsEnumerable()`) truncate the key, the descriptor re-lookup misses, and the property vanishes (`devl[{}]`). And a second, independent corruption class rides the same read: QuickJS stores WTF-8, so one lone surrogate arrives as **three** U+FFFD (the reference codec yields one).

A host-side fix exists and is validated (branch `pgp/quickjs-host-serde-fix`) — no quickjs-wasi release needed:
- `guestString()`: fast-path `toString()`, validated against the guest string's own `.length` (a plain data property, no guest code; truncation strictly shortens and the 1→3 surrogate expansion also mismatches, so equal lengths prove exactness). On mismatch, re-read via boot-captured `JSON.stringify` — QuickJS implements well-formed stringify, so NULs escape as `\u0000` and lone surrogates as `\ud800`, and host `JSON.parse` revives both byte-exactly (verified empirically for both classes).
- `shapeOf()`: enumerate string keys *inside* the VM via boot-captured `Object.keys` (same set/order as the host-side iteration it replaces), read through `guestString`.
- Verified against 3.3.0: `newString` (host→VM) and `getOwnPropertyDescriptor(string)` are length-safe — only the read direction and enumeration need this.

Parity cases added: NUL in value, NUL in object key+value, NUL in RegExp source, lone surrogate, astral pair. `nullByteWorkflow` e2e passes with the fix.

File: packages/core/src/runtime/quickjs-serde.ts [hunk 6]
```diff
@@ -0,0 +1,1996 @@
...
+    },
+    RangeError: namedErrorSubclassReducer('RangeError'),
+    ReferenceError: namedErrorSubclassReducer('ReferenceError'),
+    RetryableError: (value) => {
+      if (!isHandle(value) || !value.isError) return false;
+      if (chainedString(value, 'name') !== 'RetryableError') return false;
+      const shape = reduceErrorShape(value) as Record<string, unknown>;
+      // retryAfter is a guest Date (or string/number); normalize to an epoch
+      // timestamp exactly like the in-VM reducer.
+      let retryAfter = Date.now() + 1000;
```
Comment (pranaygp) [inline packages/core/src/runtime/quickjs-serde.ts]: **Non-deterministic value written into the event log**: the `Date.now() + 1000` fallback when `retryAfter` is absent means a replay produces different bytes than the original serialization. Everything else in the error family (cause chains, registry subclasses) round-trips correctly — this is the one spot.

File: packages/core/src/runtime/quickjs-serde.ts [hunk 7]
```diff
@@ -0,0 +1,1996 @@
...
+      registry.dispose();
+    }
+  }
+
+  // --- public API ---
+
+  return {
+    reducerKeys: Object.keys(reducers),
+    reviverKeys: Object.keys(revivers),
+    serialize(value: JSValueHandle): Uint8Array {
```
Comment (pranaygp) [inline packages/core/src/runtime/quickjs-serde.ts:2018]: **No disposal sweep after `stringify`** — reducers create handle leaves (e.g. via `dup()` in `collect`) that are never freed: roughly one leaked handle per value node, per serialize call, unbounded within a VM's lifetime. Not a crash, but with #3049's long-lived sessions it accumulates across the whole inline batch. Eager `finally` disposal (or the bulk-free arena flagged in quickjs-wasi#26) closes it.

File: packages/core/src/runtime/quickjs-runtime.ts [hunk 8]
```diff
@@ -1947,8 +1975,93 @@ function markCreated(vm: QuickJS, cidJs: string, opType?: string): void {
...
  */
-function collectDrainOperations(vm: QuickJS): PendingOperation[] {
-  using h = vm.evalCode(`(function(){
+/**
+ * Per-VM cache of serialized pending-op field bytes, keyed
+ * `correlationId:field`. A step's raw input is immutable once pushed, so
+ * its bytes are computed once even though the op is re-collected on every
+ * suspension it stays pending through.
+ */
+const pendingByteCache = new WeakMap<QuickJS, Map<string, Uint8Array>>();
```
Comment (pranaygp) [inline packages/core/src/runtime/quickjs-runtime.ts:2131]: Byte-cache notes: (1) never evicted — keys are `correlationId:field`, so a long run grows this monotonically for the VM's lifetime; (2) the suspend path passes the cache but `collectDrainOperations` doesn't, so one op can serialize twice on different paths — with any getter re-invocation that's two different byte sequences for what the log treats as one value; (3) `identities` in the serde is keyed on raw pointer and only cleared at dispose — safe today, but a future bulk-free arena reusing pointers would produce false devalue ref identity. Worth comments/guards now.

**Review by pranaygp — COMMENTED** [review]
Reviewed the incremental diff (+2823) and validated empirically against quickjs-wasi 3.3.0. Two silent data-corruption blockers on the guest→host string read — **but the good news is a host-side fix exists, is small, and is validated** (contrary to the "needs quickjs-wasi 3.4" first read): local branch `pgp/quickjs-host-serde-fix` (can push), 49 parity tests including 7 new string edge cases, full serde+runtime suites green, and `nullByteWorkflow` passes end-to-end under `WORKFLOW_VM=quickjs`. Details inline at `primitiveOf`.

Also verified/confirmed here:
- **Wire-format parity design is right** and the reducer-key exhaustiveness pinning is good; side-effect-free classification holds (boot-sampled brands, captured intrinsics, descriptor reads; the spoofed-brand test passes). One wording nit: `get` does invoke own getters (parity with the hardened node codec), so "side-effect-free" is true of *classification*, not extraction.
- **PRNG determinism on the non-snapshot path is clean** — the host-side `monotonicFactory(() => rng())` shares the one seeded instance with VM `Math.random`, and the interleaved draw sequence matches the old in-VM factory.
- **devalue 5.9.0 and quickjs-wasi 3.3.0 are stock npm, unpatched, and past the 48 h `minimumReleaseAge`** — installs are green; no dependency risk found.
- **Merge order**: `git merge-tree` vs `quickjs-vm-threshold-snapshots` reports no textual conflict, so #3048 → #3049 → this → rebase {#3250 → #3251} is mechanically safe and matches the PR body. The cost is all semantic — three silent breaks for the snapshot rebase: (1) `__generateUlid` is registered inline (the snapshot branch's own rule: host callbacks never inline — it won't be re-registered on restore); (2) the restore path's "serde survives in the heap" comment becomes false — `createQuickJSSerde(vm)` must run on restore; (3) the ULID monotonic factory's internal state is host-side now, so it's not captured by the snapshot and the `rngDraws` fast-forward lands on the wrong draw position. (The deleted `vm-bundle-entry.ts` documented the late-binding factory as the affordance for exactly this.)
- **Byte cache**: never evicted (grows for the VM's lifetime); the suspend path passes `ensurePendingByteCache(vm)` but `collectDrainOperations` passes no cache, so the same `correlationId:field` can serialize twice on different paths; and `globalThis.__rawFields` isn't cleaned up in a `finally`, so a throw mid-`dumpPendingOps` leaves it observable to workflow code.
- **Handle lifecycle**: `serialize()` has no disposal sweep after `stringify` — roughly one leaked handle per value node per call; `shapeOf` leaks the symbol-key descriptor handles; and the pointer-keyed `identities` map will produce false devalue ref-identity if the planned bulk-free arena ever reuses pointers (worth a comment now).
- Stale references: `.github/workflows/tests.yml` still lists `vm-serde-bundle.generated.ts` in the upload-artifact paths (the comment above it says to keep it aligned with `turbo.json`), and three comments in `quickjs-entrypoint.ts` still reference `globalThis[Symbol.for('workflow-serialize')]`.
- Changeset: given the wire-format work and dependency bumps, `minor` fits better than `patch` (matters on a `stable` backport even though beta numbering ignores it).

Test-hygiene notes on the (otherwise well-built) parity suite: the side-effect test installs permanent prototype patches with no restore; dead code at the byte-compare helper; and the mid-suite serde re-creation papers over cross-test handle-state coupling rather than proving it absent.

**Comment by pranaygp** [comment]
Fix branch pushed: [`pgp/quickjs-host-serde-fix`](https://github.com/vercel/workflow/compare/quickjs-host-serde...pgp/quickjs-host-serde-fix) — `guestString()` + in-VM key enumeration, covering NUL values/keys/RegExp sources and lone surrogates, with 7 new parity cases (49/49 green, `nullByteWorkflow` e2e passes). Signed.

**Comment by TooTallNate** [comment]
## Serde performance: host-side (this PR) vs in-VM bundle (base)

Benchmarked the question of whether host-side serde pays for its repeated WASM-boundary hops vs running devalue entirely inside the VM. **It's the opposite: host-side serde is faster across every payload shape tested — 1.8× to 105×.**

Method: identical script on both branches' built dists — a 10-step workflow piping a payload step→step through `runQuickJSWorkflow` full replay (fresh VM per invocation, WASM module cache warm, median of 7 after 2 warmups, same machine back-to-back). `final` = full log to completion (deserialize input + 10 step results into the VM, serialize the return out). `mid` = half log to suspension (additionally serializes the next step's input out).

| payload | serialized | in-VM final | host final | speedup | in-VM mid | host mid | speedup |
|---|---:|---:|---:|---:|---:|---:|---:|
| small | 29 B | 6.9 ms | 3.8 ms | 1.8× | 5.8 ms | 3.4 ms | 1.7× |
| wide200 (200-key obj) | 7 KB | 58.3 ms | 15.9 ms | 3.7× | 33.6 ms | 10.3 ms | 3.3× |
| array5k (5k objs) | 370 KB | 4214.9 ms | 1447.8 ms | 2.9× | 2301.4 ms | 828.6 ms | 2.8× |
| string512k | 524 KB | 2494.9 ms | **23.8 ms** | **105×** | 1362.9 ms | 15.1 ms | 90× |
| rich (Maps/Sets/Dates) | 10 KB | 129.0 ms | 63.3 ms | 2.0× | 71.5 ms | 35.7 ms | 2.0× |

Why the intuition inverts: the boundary hops are cheap C calls per value node, but the in-VM approach runs the entire devalue codec as *interpreted JS inside an interpreted VM* — QuickJS-on-WASM executes JS roughly two orders of magnitude slower than V8's JIT, and every payload byte additionally round-trips through VM-heap Uint8Arrays and in-VM UTF-8 string handling. Host-side serde does the codec work at native V8 speed and only crosses the boundary to construct/read the final value graph. The 512 KB string is the purest illustration: one `newString` handle op vs QuickJS chewing through a half-megabyte devalue parse.

Two secondary effects worth noting:
- the `small` payload's 1.8× shows the fixed win from not evaluating the VM serde bundle at every VM boot;
- unmeasured but real: the codec's temporaries no longer live in the 256 MB WASM heap, reducing VM memory pressure for large payloads.

So this PR is a performance improvement in addition to its correctness/hardening goals. Bench script available on request.

File: packages/core/src/runtime/quickjs-serde.ts (hunk 5 above)
Comment (TooTallNate) [reply packages/core/src/runtime/quickjs-serde.ts]: Fixed in f88165c1b. Confirmed empirically: `toString()` on `"ab\u0000cd"` returns `"ab"`, and NUL-bearing keys truncate in the enumeration APIs (dropping when the truncated name fails the enumerability probe, colliding when a sibling shares it). The fix funnels every guest→host string through `guestString()` — truncation is detected by comparing against `handle.length` (the true guest length, so NUL-free strings pay only a property read) and recovered via in-VM `JSON.stringify` escaping, whose output is NUL-free by construction. Key enumeration verifies the fast host-string path against a guest `Object.keys` count + duplicate check and re-extracts through length-aware key handles on mismatch; `get`/`hasOwn` route NUL-bearing keys through handle-keyed access (`vm.newString` is length-aware, verified). The parse/build side was already safe (`define()` goes through guest key handles). Regression tests cover values (leading/trailing/middle/multi NUL), byte parity with the reference codec, and both enumeration corruption shapes; `nullByteWorkflow` passes under `WORKFLOW_VM=quickjs` locally.

File: packages/core/src/runtime/quickjs-serde.ts (hunk 6 above)
Comment (TooTallNate) [reply packages/core/src/runtime/quickjs-serde.ts]: Fixed in f88165c1b — the fallback now reads the GUEST clock (a captured `Date.now`, which sits on the deterministic replay clock at the WASI layer) instead of the host wall clock. The in-VM reducer's `Date.now()` was replay-stable by construction; the host port silently swapped it for wall time — this restores the original semantics rather than changing the wire shape (node's reducer always emits a number here, so omitting the field would have been a cross-engine format change).

File: packages/core/src/runtime/quickjs-serde.ts (hunk 7 above)
Comment (TooTallNate) [reply packages/core/src/runtime/quickjs-serde.ts:2018]: Fixed in f88165c1b — serialize and deserialize now sweep every intermediate handle their pass creates (call/invoke results, descriptor reads, dups, parse-op constructions; deserialize escapes the root). One finding worth flagging: the obvious implementation — `vm.withScope` — is UNSAFE here. The library scope registers every handle constructed while active, including the ones `handleHostCall` wraps around C-owned argv pointers when a host callback runs (our Map/Set/Headers forEach visitors, mid-pass). Disposing those double-frees guest values; observed as WASM `memory access out of bounds` in the parity suite. The sweep therefore uses module-owned tracking fed only by this module's creation funnels, which is safe by construction. Filed as a follow-up note on the PR for an upstream fix (suspend the active scope in the trampoline, as `getPromiseThen` already does). `identities` is also cleared per pass now — with handles being freed, pointer reuse across passes could otherwise alias stale identity entries (your point 3 on the byte-cache thread).

File: packages/core/src/runtime/quickjs-runtime.ts (hunk 8 above)
Comment (TooTallNate) [reply packages/core/src/runtime/quickjs-runtime.ts:2131]: Fixed in f88165c1b — all three points: (1) eviction — the collection pass now surfaces settled ops (created + resolver-less + no abort in flight, which neither the suspension nor the drain filter can ever match again) in the same guest sweep, and their cache entries are dropped, bounding the cache by the live pending set; (2) `collectDrainOperations` now shares the per-VM cache, so an op serialized on the suspension path reuses those exact bytes at terminal drain instead of risking a second, getter-divergent byte sequence; (3) `identities` is now cleared per serialize pass (it only needs intra-pass stability), which makes it immune to pointer reuse — necessary anyway now that the pass-disposal sweep actually frees handles (see the `:1967` thread).

File: packages/core/src/runtime/quickjs-serde.ts (hunk 7 above)
Comment (TooTallNate) [reply packages/core/src/runtime/quickjs-serde.ts:2018]: Follow-up: the library-side fix is now up as vercel-labs/quickjs-wasi#31 — host-callback `this`/argument handles become 'borrowed' (never scope-tracked, `dispose()` no-op, `dup()` for retention), with regression tests covering the exact forEach-visitor corruption shape found here. Once that ships, `vm.withScope` becomes a valid alternative to this PR's module-owned tracking; the tracking approach stays correct either way (it's a strict subset — only handles this module creates), so no change needed here.

File: packages/core/src/runtime/quickjs-serde.ts [hunk 9]
```diff
@@ -0,0 +1,2118 @@
...
+   * so a mismatch against the extracted string's length means data was
+   * lost. The recovery path escapes the string INSIDE the VM with the
+   * captured `JSON.stringify` (the escaped form is NUL-free by
+   * construction, so its own extraction cannot truncate) and parses it
+   * host-side. NUL-free strings — the overwhelming majority — pay only
+   * the `.length` read.
+   */
+  const guestString = (handle: JSValueHandle): string => {
+    const fast = handle.toString();
+    if (fast.length === handle.length) return fast;
```
Comment (karthikscale3) [inline packages/core/src/runtime/quickjs-serde.ts]: **[P1 blocker] The length equality is not sufficient to prove that `JS_ToCString` was lossless because its two known corruptions can cancel each other out.** I reproduced this with a focused parity test using the guest string "\ud800\u0000a": the lone surrogate expands to three U+FFFD characters while the NUL truncates the two-code-unit suffix, so both `fast.length` and `handle.length` are 3 and this returns the corrupt fast value. The durable bytes were `devl["���"]`; the reference codec produced `devl["�\u0000a"]`. All existing 49 serde tests passed while this added case failed. Please make the fallback trigger on length mismatch **or** a U+FFFD in the fast result (legitimate replacement characters can safely take the slow path), and add mixed lone-surrogate+NUL cases for values and keys.

File: packages/core/src/runtime/quickjs-serde.ts [hunk 10]
```diff
@@ -0,0 +1,2118 @@
...
+
+const encoder = new TextEncoder();
+const decoder = new TextDecoder();
+const FORMAT_PREFIX_LENGTH = 4;
+
+// ---- base64 (host-side; wire-compatible with the old in-VM btoa path) ----
+
+function bytesToBase64(bytes: Uint8Array): string {
+  if (bytes.length === 0) return '.';
+  return Buffer.from(bytes.buffer, bytes.byteOffset, bytes.byteLength).toString(
```
Comment (karthikscale3) [inline packages/core/src/runtime/quickjs-serde.ts]: **[P2; blocker if WASM-only/non-Node hosts are in scope] This adds an unconditional Node `Buffer` dependency to every binary serde path.** In Cloudflare Workers and similar runtimes, typed-array/ArrayBuffer serialization now throws unless the bundler injects a Buffer polyfill. Main already has one `Buffer.isBuffer()` diagnostic on QuickJS replay, so there is existing portability debt, but this PR expands it into the codec itself and conflicts with the QuickJS engine's WASM-only portability goal. Please use the portable strategy used elsewhere: native Uint8Array base64 methods when available, Buffer only when Node is detected, and `btoa`/`atob` fallback.

**Review by karthikscale3 — COMMENTED** [review]
Focused review: two inline findings—one reproducible serialization correctness blocker and one portability concern.

File: packages/core/src/runtime/quickjs-serde.ts (hunk 9 above)
Comment (TooTallNate) [reply packages/core/src/runtime/quickjs-serde.ts]: Excellent catch — fixed in d101fa25b, and your reproduction understated it slightly: probing this build showed a BARE lone surrogate corrupts with matching lengths (1:1 U+FFFD replacement, no NUL needed), and the JSON.stringify slow path was itself lossy for lone surrogates — QuickJS passes them through raw, and the C-string extraction of *its* output corrupts them. So the fix is two-part: (1) the fast value is accepted only on length match AND no U+FFFD (legit-U+FFFD strings take the loss-free slow path, per your suggestion); (2) the slow path now escapes INSIDE the VM to printable ASCII via a new captured `escapeString` intrinsic (per-code-unit \\uXXXX on WTF-16, so lone surrogates survive) and JSON-parses host-side. The key-enumeration guard gains the same U+FFFD scan (lone-surrogate keys corrupt with count and uniqueness intact), and `get`/`hasOwn` route handle-keyed for keys carrying a NUL or UNPAIRED surrogate (`vm.newString` verified WTF-16-preserving; paired surrogates encode fine). Tests added: your exact canceling case, bare surrogates, legit-U+FFFD passthrough, reference-codec byte parity, and surrogate/mixed keys.

File: packages/core/src/runtime/quickjs-serde.ts (hunk 10 above)
Comment (TooTallNate) [reply packages/core/src/runtime/quickjs-serde.ts]: Fixed in d101fa25b — the base64 helpers are now feature-detected exactly as you suggested: `Uint8Array.fromBase64`/`toBase64` when available, `Buffer` when present, `btoa`/`atob` loop otherwise. No unconditional Node dependency remains in the codec (the one pre-existing `Buffer.isBuffer` diagnostic on the replay path is debug-log-only and untouched here). The wire-parity suite pins that all three paths produce identical bytes on this runtime's active path.

**Review by karthikscale3 — APPROVED** [review]
ai review: Re-checked the latest head, lossless QuickJS string transport, portable base64 fallbacks, regression coverage, and current CI. No remaining code-review blockers found.
